core/server/master/src/main/java/alluxio/master/file/meta/MountTable.java

```diff
@@ -104,11 +107,60 @@ public MountTable(UfsManager ufsManager, MountInfo rootMountInfo) {
    * @throws InvalidPathException if an invalid path is encountered
    */
   public void add(Supplier<JournalContext> journalContext, AlluxioURI alluxioUri, AlluxioURI ufsUri,
-      long mountId, MountPOptions options) throws FileAlreadyExistsException, InvalidPathException {
+      long mountId, MountPOptions options) throws FileAlreadyExistsException, InvalidPathException,
+      IOException {
+    // validate the Mount operation first, error will be thrown if the operation is invalid
+    try (ValidatedPathPair validatedPathPair = validateMountPoint(alluxioUri, ufsUri)) {
+      add(journalContext, validatedPathPair, mountId, options);
+    }
+  }
+
+  /**
+   * Mounts based on the given ValidatedPathPair. It will directly insert a new entry into
+   * MountTable via the mState. When calling this add, the validatedPathPair must hold the write
+   * lock resource.
+   * @param journalContext the journal context
+   * @param validatedPathPair an Alluxio path URI
+   * @param mountId the mount id
+   * @param options the mount options
+   */
+  public void add(Supplier<JournalContext> journalContext, ValidatedPathPair validatedPathPair,
+                  long mountId, MountPOptions options) {
+    AlluxioURI alluxioUri = validatedPathPair.getAlluxioPath();
+    AlluxioURI ufsUri = validatedPathPair.getUfsPath();
     String alluxioPath = alluxioUri.getPath().isEmpty() ? "/" : alluxioUri.getPath();
     LOG.info("Mounting {} at {}", ufsUri, alluxioPath);
 
     try (LockResource r = new LockResource(mWriteLock)) {
+      Map<String, String> properties = options.getPropertiesMap();
+      mState.applyAndJournal(journalContext, AddMountPointEntry.newBuilder()
+          .addAllProperties(properties.entrySet().stream()
+              .map(entry -> StringPairEntry.newBuilder()
+                  .setKey(entry.getKey()).setValue(entry.getValue()).build())
+              .collect(Collectors.toList()))
+          .setAlluxioPath(alluxioPath)
+          .setMountId(mountId)
+          .setReadOnly(options.getReadOnly())
+          .setShared(options.getShared())
+          .setUfsPath(ufsUri.toString())
+          .build());
+    }
+  }
+
+  /**
+   * Verify if the given (alluxioPath, ufsPath) can be inserted into MountTable.
+   * @param alluxioUri the alluxio path that is about to be the mountpoint
+   * @param ufsUri the UFS path that is about to mount
+   * @return the (alluxioPath, ufsPath) that is validated as a legal mount entry
+   */
+  public ValidatedPathPair validateMountPoint(AlluxioURI alluxioUri, AlluxioURI ufsUri)
```

Comment:
Instead of locking inside the `validateMountPoint()` and passing it to `add()` (because that caller also needs to lock), do it outside like
```
mMountTable.writeLock();
try {
  validateMountPoint();
  mMountTable.add();
} finally {
  mountTableLock.writeUnlock();
}
```

An alternative is to use try-with-resource: `try(LockResource lr = new LockResource(mMountTable.getWriteLock())) {...}`

This way you always put the locking and unlocking together. This pattern is much easier to follow and hard to miss. Tracing a lock object here and there is error-prune. Don't forget to write unit tests like adding multiple mount points to the MountTable, to validate your locks are released correctly in corner cases.